### DISCUSSION TITLE
How can I extend OpenTelemetry Java with a new "digest" instrument?

### DISCUSSION THREAD

**Discussion by pedrolamarao** [open-telemetry/opentelemetry-java#5186] [Q&A]
In my understanding of the ticket below, "summaries" were removed from OpenTelemetry Metrics before the initial stable release over concerns about the summability of summary deltas.
https://github.com/open-telemetry/opentelemetry-proto/pull/199

T-Digests are a distributional sketch capable of maintaining high accuracy of quantiles that also allows merging while maintaining accuracy.
https://github.com/tdunning/t-digest

I have raised a discussion point on the possibility of applying T-Digests to achieve "summable summaries" in the OpenTelemetry Specification repository:
https://github.com/open-telemetry/opentelemetry-specification/discussions/3118

I would like to experiment adding such an instrument to try and advance this idea to the prototype phase. I am very confortable writing in Java and other languages for the JVM. Could someone point me in the direction of extending OpenTelemetry Java to add new experimental instruments?

**Comment by jkwatson** [comment]
I don't think you want a new instrument for this, but a new aggregation for the Histogram instrument, correct?

**Reply by pedrolamarao** [reply]
> I don't think you want a new instrument for this, but a new aggregation for the Histogram instrument, correct?

Not really.
But I do not yet understand T-Digest serialization to be able to explain what is the difference.
Ted Dunning has expressed interest in how OpenTelemetry could apply T-Digests, let me ping him.

**Reply by jkwatson** [reply]
We definitely do _not_ want to add a new Instrument. The Histogram instrument records values that can be turned into t-digest information, correct? That's the aggregation layer, not a part of the metric API, which is where the instruments live.

If you want to try a new experimental aggregation, take a look at how the exponential histogram aggregation was implemented originally. I believe it was in an `internal` package, and thus marked as unstable and unsupported.

**Reply by tdunning** [reply]
I am symmetrically ignorant about what what the Histogram instrument does exactly and what it expects in terms of serialization.

The t-digest is a sketch of the data added to it so far. Such a digest is strictly bounded in size and can be serialized and saved. It can later be combined with other digests that summarize disjoint data to get estimates of the distribution of the values added to both sketches. This combination can be done essentially to any degree desired.

What does the Histogram instrument expect?

**Reply by jkwatson** [reply]
The histogram instrument only has methods to record values. The implementation of that is provided by the aggregation, which lives in  the SDK. Currently, a histogram can be aggregated with a number of different aggregations, including both fixed-bucket histograms and base-2-exponential histograms. 

I believe what you want is to add a new aggregation option for the t-digest. Does OTEL have a way to export t-digest data via OTLP? If not, that seems like the first step, which will require work in the OTLP specification.

**Reply by tdunning** [reply]
OK. Adding values is a piece of cake.

What are OTEL and OTLP?

**Reply by jkwatson** [reply]
Otel is OpenTelemetry. OTLP is the OpenTelemetry protocol, the standard wire format for otel. 

I suggest you spend some time with the otel specification before tackling something as complex as adding an entirely new new aggregation.

**Reply by pedrolamarao** [reply]
I am familiar with OTEL and OTLP and indend to support this work. If I understand the conversation so far correctly, it should be possible to define "buckets" dinamically to corresponde with T-Digest's "centroids". I originally assumed that the histogram instrument required "buckets" with fixed, predefined size.

For public reference, please correct me Ted!, this is the most current published article explaining T-Digest: https://www.sciencedirect.com/science/article/pii/S2665963820300403

**Reply by jkwatson** [reply]
I recommend getting familiar with how the other aggregations for the histogram instrument work. Hopefully that will clear things up a bit about what this work might entail. 

Here is the implementation of explicit buckets: https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/ExplicitBucketHistogramAggregation.java

Here is the implementation of exponential buckets: https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/Base2ExponentialHistogramAggregation.java

The main question you'll have to answer is how to convert your t-digest things into metric points that are exportable. And whether we even have appropriate point implementations at this point. And, then, how you might export this data somewhere so it could be used by a backend that supported this shape of data.

**Comment by tdunning** [comment]
John,

Thanks for answering my questions.

T-digest can produce any percentiles of interest. Commonly, these include
median (0.5 quantile) and various tail measurements (0.9, 0.95, 0.99,
0.999, 0.9999) for instance. You can specify which quantiles at any time.

**Comment by pedrolamarao** [comment]
This week I have time to dedicate to this activity. I'm having trouble building both `main` and `release/1.23.x`. I've filed #5250 about this. If someone can give me guidance on how to build opentelemetry-java, please advise.

**Comment by pedrolamarao** [comment]
In order to help our conversation move forward, I have implemented a new type of aggregator, the `TDigestAggregator`, that you can find here: https://github.com/pedrolamarao/opentelemetry-java/blob/digests/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/aggregator/TDigestAggregator.java

This aggregator records data by feeding it to a `TDigest`. It generates a `HistogramDataPoint` by transforming each TDigest centroid into a histogram datum by assigning centroid mean to histogram boundary and centroid count to histogram count. But this seems conceptually wrong: a TDigest centroid mean is not a boundary.

**Reply by jkwatson** [reply]
Yes, this is where I was wondering if OTLP has the right data structures currently to support tdigest export. You might need to go to the spec and work with the technical committee (paging @jack-berg ) to figure out how to extend OTLP to support tdigest structures.

**Reply by pedrolamarao** [reply]
I'll attempt to understand this from the perspective of the collector, building an experimental collector capable of understanding this data.

**Reply by jack-berg** [reply]
One of the reasons that custom aggregators aren't supported (at least yet) is that all aggregators have to output metric points as defined in the [metric data model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#metric-points). Point kinds include: sum, gauge, histogram, exponential histogram, and summary (legacy). There are only so many ways to usefully aggregate measurements into these metric points, and most are covered by the small set of built in aggregators.

I'm not familiar with T-Digest, so I don't have intuition about the information needs to be encoded to be sent over the wire. I assume that in order to be summable, certain data has to be present. You'll need to study the data model (perhaps take a look at the [metric proto](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto) definitions to for a concrete description) to determine if its feasible to encode T-Digest data into one of the existing metric point kinds without violating the semantics of the point fields. If its not possible, you'll have to propose extending the data model.

To extend the data model, you'll have to open an issue in the specification describing your proposal and convince others of its merit. I see you've opened a discussion - you'll want to file an issue instead. Extending the metric data model is a significant enough change that it requires going through the [OTEP](https://github.com/open-telemetry/oteps) process. However, I suggest starting with an issue, and perhaps attending and presenting the idea in the [specification SIG](https://github.com/open-telemetry/community#specification-sigs), which meets each Tuesday at 8am PST.

**Comment by tdunning** [comment]
You can estimate boundaries between centroids based on their counts.

I would not recommend serializing a t-digest with boundaries, but they can
be used for visualization just fine.

You can also use the centroids as boundaries and assign half of the points
to either side unless the count = 1. For unit count, there is a singleton
point at the centroid so you should handle that with a special case.

**Reply by pedrolamarao** [reply]
I think the end game is to reconstruct the TDigest at the point of collection, enabling the collector to sum TDigest "deltas". This would require the collector to understand the digest semantics. I'll try to demonstrate an experimental collector capable of doing this.

**Comment by tdunning** [comment]
OK.

Take a digest where the centroid positions are x_i and the centroid weights
are c_i where i = 1..n. Note that c_1 = c_n = 1.

You can encode this digest as a histogram in a couple of ways that will
look kind of OK when viewed as a histogram and which allow perfect
reconstruction of the histogram (if you know the compression factor
separately).

If you can only have integer weights in the histogram, then use x_i as the
bounds of n-1 bins and take the histogram weights as

h_1 = c_1 + c_2 = c_2 + 1
h_j = c_{j+1} where j = 1..n-1.

Reconstruction is trivial.

If you can have non-integer weights in the histogram and want ever so
slightly better looking histograms, take x_i as the bounds of n-1 bins and
use weights

h_1 = 1 + c_2/2
h_j = c_j/2 + c_{j+1}/2       where j = 2 ... n-2
h_{n-1} = c_{n-1}/2 + 1

Reconstruction is only slightly less trivial.

**Reply by pedrolamarao** [reply]
I'll put some time to try and demonstrate all alternatives for everyone's consideration.
